### PR TITLE
feat: Add grpc gateway annotations

### PR DESCRIPTION
**Pull request by aleem1314**
closes: #182

### CODE REVIEW

File: proto/regen/data/v1alpha1/query.proto [hunk 1]
```diff
@@ -11,10 +12,14 @@ option go_package = "github.com/regen-network/regen-ledger/x/data";
 service Query {
 
     // ByCid queries data based on its CID.
-    rpc ByCid(QueryByCidRequest) returns (QueryByCidResponse);
+    rpc ByCid(QueryByCidRequest) returns (QueryByCidResponse) {
+        option (google.api.http).get = "/regen/data/v1alpha1/cids";
```
Comment (amaurymartiny) [inline proto/regen/data/v1alpha1/query.proto:16]:
```suggestion
        option (google.api.http).get = "/regen/data/v1alpha1/cids/{cid}";
```

Comment (amaurymartiny) [reply proto/regen/data/v1alpha1/query.proto:16]: We need to do that, right? Most of the annotations lack this

File: third_party/proto/google/protobuf/any.proto [hunk 2]
```diff
@@ -0,0 +1,161 @@
+// Protocol Buffers - Google's data interchange format
```
Comment (aaronc) [inline third_party/proto/google/protobuf/any.proto:1]: We don't need this file

File: proto/regen/ecocredit/v1alpha1/query.proto [hunk 3]
```diff
@@ -10,20 +10,30 @@ option go_package = "github.com/regen-network/regen-ledger/x/ecocredit";
 // Msg is the regen.ecocredit.v1alpha1 Query service.
 service Query {
     // ClassInfo queries for information on a credit class.
-    rpc ClassInfo(QueryClassInfoRequest) returns (QueryClassInfoResponse);
+    rpc ClassInfo(QueryClassInfoRequest) returns (QueryClassInfoResponse) {
+        option (google.api.http).get = "/regen/ecocredit/v1alpha1/classinfo/{class_id}";
```
Comment (aaronc) [inline proto/regen/ecocredit/v1alpha1/query.proto:14]:
```suggestion
        option (google.api.http).get = "/regen/ecocredit/v1alpha1/class_info/{class_id}";
```

File: proto/regen/ecocredit/v1alpha1/query.proto [hunk 4]
```diff
@@ -10,20 +10,30 @@ option go_package = "github.com/regen-network/regen-ledger/x/ecocredit";
...
+    }
 
     // BatchInfo queries for information on a credit batch.
-    rpc BatchInfo(QueryBatchInfoRequest) returns (QueryBatchInfoResponse);
+    rpc BatchInfo(QueryBatchInfoRequest) returns (QueryBatchInfoResponse){
+        option (google.api.http).get = "/regen/ecocredit/v1alpha1/batchinfo/{batch_denom}";
```
Comment (aaronc) [inline proto/regen/ecocredit/v1alpha1/query.proto:19]:
```suggestion
        option (google.api.http).get = "/regen/ecocredit/v1alpha1/batch_info/{batch_denom}";
```

File: proto/regen/data/v1alpha1/query.proto [hunk 5]
```diff
@@ -11,10 +12,14 @@ option go_package = "github.com/regen-network/regen-ledger/x/data";
 service Query {
 
     // ByCid queries data based on its CID.
-    rpc ByCid(QueryByCidRequest) returns (QueryByCidResponse);
+    rpc ByCid(QueryByCidRequest) returns (QueryByCidResponse) {
+        option (google.api.http).get = "/regen/data/v1alpha1/cids/{cid}";
```
Comment (aaronc) [inline proto/regen/data/v1alpha1/query.proto:16]:
```suggestion
        option (google.api.http).get = "/regen/data/v1alpha1/{cid}";
```

File: proto/regen/data/v1alpha1/query.proto [hunk 6]
```diff
@@ -11,10 +12,14 @@ option go_package = "github.com/regen-network/regen-ledger/x/data";
...
+    }
 
     // BySigner queries data based on signers.
-    rpc BySigner(QueryBySignerRequest) returns (QueryBySignerResponse);
+    rpc BySigner(QueryBySignerRequest) returns (QueryBySignerResponse){
+        option (google.api.http).get = "/regen/data/v1alpha1/signers/{signer}";
```
Comment (aaronc) [inline proto/regen/data/v1alpha1/query.proto:21]:
```suggestion
        option (google.api.http).get = "/regen/data/v1alpha1/by-signer/{signer}";
```

**Review by aaronc — CHANGES_REQUESTED** [review]

**Review by amaurymartiny — APPROVED** [review]
would be cool to have 1 or 2 integration tests too, for a later PR

File: proto/regen/data/v1alpha1/query.proto [hunk 7]
```diff
@@ -11,10 +12,14 @@ option go_package = "github.com/regen-network/regen-ledger/x/data";
 service Query {
 
     // ByCid queries data based on its CID.
-    rpc ByCid(QueryByCidRequest) returns (QueryByCidResponse);
+    rpc ByCid(QueryByCidRequest) returns (QueryByCidResponse) {
+        option (google.api.http).get = "/regen/data/v1alpha1/{cid}";
```
Comment (anilCSE) [inline proto/regen/data/v1alpha1/query.proto:16]:
```suggestion
        option (google.api.http).get = "/regen/data/v1alpha1/cids/{cid}";
```

File: proto/regen/ecocredit/v1alpha1/query.proto [hunk 8]
```diff
@@ -10,20 +10,30 @@ option go_package = "github.com/regen-network/regen-ledger/x/ecocredit";
...
     // BatchInfo queries for information on a credit batch.
-    rpc BatchInfo(QueryBatchInfoRequest) returns (QueryBatchInfoResponse);
+    rpc BatchInfo(QueryBatchInfoRequest) returns (QueryBatchInfoResponse){
+        option (google.api.http).get = "/regen/ecocredit/v1alpha1/batch_info/{batch_denom}";
+    }
 
     // Balance queries the balance (both tradable and retired) of a given credit batch for a given account.
-    rpc Balance(QueryBalanceRequest) returns (QueryBalanceResponse);
+    rpc Balance(QueryBalanceRequest) returns (QueryBalanceResponse) {
+        option (google.api.http).get = "/regen/ecocredit/v1alpha1/{batch_denom}/balances/{account}";
```
Comment (anilCSE) [inline proto/regen/ecocredit/v1alpha1/query.proto:24]:
```suggestion
        option (google.api.http).get = "/regen/ecocredit/v1alpha1/balances/{account}/{batch_denom}";
```
or 
```suggestion
        option (google.api.http).get = "/regen/ecocredit/v1alpha1/batch_info/{batch_denom}/balances/{account}";
```
or 
```suggestion
        option (google.api.http).get = "/regen/ecocredit/v1alpha1/batches/{batch_denom}/balances/{account}";
```

File: proto/regen/ecocredit/v1alpha1/query.proto [hunk 9]
```diff
@@ -10,20 +10,30 @@ option go_package = "github.com/regen-network/regen-ledger/x/ecocredit";
...
+    }
 
     // Supply queries the tradable and retired supply of a credit batch.
-    rpc Supply(QuerySupplyRequest) returns (QuerySupplyResponse);
+    rpc Supply(QuerySupplyRequest) returns (QuerySupplyResponse) {
+        option (google.api.http).get = "/regen/ecocredit/v1alpha1/supply/{batch_denom}";
```
Comment (anilCSE) [inline proto/regen/ecocredit/v1alpha1/query.proto:29]: @amaurymartiny is `batches` a correct usage here?
```suggestion
        option (google.api.http).get = "/regen/ecocredit/v1alpha1/batches/{batch_denom}/supply";
```

Comment (anilCSE) [reply proto/regen/ecocredit/v1alpha1/query.proto:29]: if not, lets use `batch_info/{batch_denom}/supply`

File: proto/regen/ecocredit/v1alpha1/query.proto [hunk 10]
```diff
@@ -10,20 +10,30 @@ option go_package = "github.com/regen-network/regen-ledger/x/ecocredit";
...
+    }
 
     // Precision queries the number of decimal places that can be used to represent credit batch units.
     // See Tx/SetPrecision for more details.
-    rpc Precision(QueryPrecisionRequest) returns (QueryPrecisionResponse);
+    rpc Precision(QueryPrecisionRequest) returns (QueryPrecisionResponse) {
+        option (google.api.http).get = "/regen/ecocredit/v1alpha1/precision/{batch_denom}";
```
Comment (anilCSE) [inline proto/regen/ecocredit/v1alpha1/query.proto:35]:
```suggestion
        option (google.api.http).get = "/regen/ecocredit/v1alpha1/batches/{batch_denom}/precision";
```

**Review by anilCSE — CHANGES_REQUESTED** [review]

**Comment by blushi** [comment]
@aleem1314 Now that the group module is merged, we will also need to add grpc gateway annotations to it.

**Comment by amaurym** [comment]
It seems that this fails because of the custom code generator. Maybe we should fix it in the SDK first? ref: https://github.com/cosmos/cosmos-sdk/pull/9302

File: scripts/protocgen.sh [hunk 11]
```diff
@@ -15,6 +15,10 @@ protoc_gen_doc() {
   go get -u github.com/pseudomuto/protoc-gen-doc/cmd/protoc-gen-doc 2>/dev/null
 }
 
+protoc_gen_gateway2() {
+  go get github.com/aleem1314/grpc-gateway/protoc-gen-grpc-gateway2@v1.16.5-final 2>/dev/null
```
Comment (amaurym) [inline scripts/protocgen.sh:19]: Thanks @aleem1314 for taking a look! I'm not sure it's worth to fork grpc-gateway. Maybe it's better to just remove our customer codegen for now.

**Review by amaurym — APPROVED** [review]
Yes, this is going in the right direction (unfortunately for now). Pre-approving.

We can also remove the `protoc-gen-go-cosmos2` folder.

File: .codecov.yml [hunk 12]
```diff
@@ -26,4 +26,6 @@ ignore:
   - "build"
   - "tests/*"
   - "contrib"
-  - "**/*.pb.go"
\ No newline at end of file
+  - "**/*.pb.go"
+  - "**/*.pb.gw.go"
+  - "scripts"
```
Comment (anilCSE) [inline .codecov.yml:31]:
```suggestion
  - "scripts"
  
```

File: proto/regen/data/v1alpha2/query.proto [hunk 13]
```diff
@@ -11,10 +12,14 @@ option go_package = "github.com/regen-network/regen-ledger/x/data";
 // Query is the regen.data.v1alpha1 Query service
 service Query {
   // ByHash queries data based on its ContentHash.
-  rpc ByHash (QueryByHashRequest) returns (QueryByHashResponse);
+  rpc ByHash (QueryByHashRequest) returns (QueryByHashResponse) {
+    option (google.api.http).get = "/regen/data/v1alpha2/by-hash";
```
Comment (anilCSE) [inline proto/regen/data/v1alpha2/query.proto:16]: why are we using `v1alpha2` here? If we stick to this, update the documentation above.

File: proto/regen/ecocredit/v1alpha1/query.proto [hunk 14]
```diff
@@ -1,30 +1,39 @@
...
-
-  // Supply queries the tradable and retired supply of a credit batch.
-  rpc Supply(QuerySupplyRequest) returns (QuerySupplyResponse);
-
-  // Precision queries the number of decimal places that can be used to
-  // represent credit batch units. See Tx/SetPrecision for more details.
-  rpc Precision(QueryPrecisionRequest) returns (QueryPrecisionResponse);
+    // ClassInfo queries for information on a credit class.
+    rpc ClassInfo(QueryClassInfoRequest) returns (QueryClassInfoResponse) {
+        option (google.api.http).get = "/regen/ecocredit/v1alpha1/class_info/{class_id}";
```
Comment (anilCSE) [inline proto/regen/ecocredit/v1alpha1/query.proto:14]: let's use either `-` or `_`. Looks like we are using both for word separation

**Review by anilCSE — APPROVED** [review]

**Review by aaronc — APPROVED** [review]

**Pull request merged**